kubelet/stats: deduplicate makePodStorageStats

haircommander opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind cleanup
#### What this PR does / why we need it:
Deduplicates pod storage stats generation between CRI and cadvisor stats providers, and unifies their behavior. This fixes a bug where cadvisor stats provider excessively reported a GC'd directory as not found

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/106957

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

haircommander committed:
kubelet/stats: deduplicate makePodStorageStats

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander commented:
/sig node

ehashman commented:
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2

haircommander commented:
/retest

ehashman commented:
/priority important-longterm
/triage accepted

ehashman commented on pkg/kubelet/stats/helper.go:
```diff
@@ -433,3 +435,29 @@ func addUsage(first, second *uint64) *uint64 {
 	return &total
 }
+
+func makePodStorageStats(s *statsapi.PodStats, rootFsInfo *cadvisorapiv2.FsInfo, resourceAnalyzer stats.ResourceAnalyzer, hostStatsProvider HostStatsProvider, isCRIStatsProvider bool) {
```
Diff:

```diff
diff --git a/home/ehashman/tmp/foo2 b/home/ehashman/tmp/foo1
index 4ba09d2bead..2dc73b0ed5d 100644
--- a/home/ehashman/tmp/foo2
+++ b/home/ehashman/tmp/foo1
@@ -1,12 +1,12 @@
-func (p *criStatsProvider) makePodStorageStats(s *statsapi.PodStats, rootFsInfo *cadvisorapiv2.FsInfo) {
+func makePodStorageStats(s *statsapi.PodStats, rootFsInfo *cadvisorapiv2.FsInfo, resourceAnalyzer stats.ResourceAnalyzer, hostStatsProvider HostStatsProvider, isCRIStatsProvider bool) {
        podNs := s.PodRef.Namespace
        podName := s.PodRef.Name
        podUID := types.UID(s.PodRef.UID)
-       vstats, found := p.resourceAnalyzer.GetPodVolumeStats(podUID)
+       vstats, found := resourceAnalyzer.GetPodVolumeStats(podUID)
        if !found {
                return
        }
-       logStats, err := p.hostStatsProvider.getPodLogStats(podNs, podName, podUID, rootFsInfo)
+       logStats, err := hostStatsProvider.getPodLogStats(podNs, podName, podUID, rootFsInfo)
        if err != nil {
                klog.ErrorS(err, "Unable to fetch pod log stats", "pod", klog.KRef(podNs, podName))
                // If people do in-place upgrade, there might be pods still using
@@ -14,12 +14,12 @@ func (p *criStatsProvider) makePodStorageStats(s *statsapi.PodStats, rootFsInfo
                // We should continue generating other stats in that case.
                // calcEphemeralStorage tolerants logStats == nil.
        }
-       etcHostsStats, err := p.hostStatsProvider.getPodEtcHostsStats(podUID, rootFsInfo)
+       etcHostsStats, err := hostStatsProvider.getPodEtcHostsStats(podUID, rootFsInfo)
        if err != nil {
                klog.ErrorS(err, "Unable to fetch pod etc hosts stats", "pod", klog.KRef(podNs, podName))
        }
        ephemeralStats := make([]statsapi.VolumeStats, len(vstats.EphemeralVolumes))
        copy(ephemeralStats, vstats.EphemeralVolumes)
        s.VolumeStats = append(append([]statsapi.VolumeStats{}, vstats.EphemeralVolumes...), vstats.PersistentVolumes...)
-       s.EphemeralStorage = calcEphemeralStorage(s.Containers, ephemeralStats, rootFsInfo, logStats, etcHostsStats, true)
+       s.EphemeralStorage = calcEphemeralStorage(s.Containers, ephemeralStats, rootFsInfo, logStats, etcHostsStats, isCRIStatsProvider)
 }
```

ehashman commented on pkg/kubelet/stats/cadvisor_stats_provider.go:
```diff
@@ -132,25 +132,9 @@ func (p *cadvisorStatsProvider) ListPodStats() ([]statsapi.PodStats, error) {
-		// Lookup the volume stats for each pod.
-		podUID := types.UID(podStats.PodRef.UID)
-		var ephemeralStats []statsapi.VolumeStats
-		if vstats, found := p.resourceAnalyzer.GetPodVolumeStats(podUID); found {
```
There is a semantic difference here now. Previously (in the deleted code), if we couldn't find volume stats, we keep looking for log/etchosts stats. Now, we return if the vstats aren't found. Does this have any unintentional effects?

haircommander replied:
I would venture to guess not, considering CRI stats provider has the same behavior with no known issues. @dashpole do you have any insight here?

dashpole replied:
It shouldn't matter too much, although I would prefer the previous behavior (continuing if we don't find volume stats).

haircommander replied:
I believe not skipping is why we're seeing the logs in [this](https://github.com/kubernetes/kubernetes/issues/106957) issue. I would be open to continuing with stats gathering, but maybe make failures silent rather than erroring? To me, it makes sense to skip if we can't find volume stats, though.

rpthms commented:
@haircommander Any chance you could take a look at the comments raised in this PR if you got time to spare? Thanks!

haircommander commented:
/retest

haircommander commented:
I personally think this route still makes sense. the behavior change is consistent with the cri stats manager, and makes logical sense to me. I am open to more feedback though

aneagoe commented:
@andrewsykim @tallclair any chance you have some time to review this? It would be great to see some progress on this issue. Thanks!

haircommander commented:
/assign @endocrimes @mrunalp @rphillips @saschagrunert 

Anyone have any feedback or can we move forward with this

rphillips commented on pkg/kubelet/stats/helper.go:
```diff
@@ -433,3 +435,29 @@ func addUsage(first, second *uint64) *uint64 {
+	podUID := types.UID(s.PodRef.UID)
+	vstats, found := resourceAnalyzer.GetPodVolumeStats(podUID)
+	if !found {
+		return
```
I don't think this function should require Volume stats to be present... the old code didn't require them either.

haircommander replied:
why not? CRI stats code did, and seems to be working fine https://github.com/kubernetes/kubernetes/pull/108855/files#diff-27608c014abc817b90c50f8bbca712d16caa731c5d12465435cde848c72c2dc9L483

haircommander committed:
kubelet/stats: drop makePodStorageStats errors to V(6)

and by doing so, fix a bug where the stats providers report a directory is not found after a pod's storage is removed

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander commented:
Alright, I've updated to:
adopt the cadvisor stats provider behavior where we continue the stats collection, even if volume stats are not present
reduced the verbosity of the logs by increasing to 6

@dashpole @rphillips PTAL

dchen1107 commented:
/lgtm

endocrimes reviewed (approved):
/lgtm

dims commented:
/milestone v1.26

cc @mrunalp
